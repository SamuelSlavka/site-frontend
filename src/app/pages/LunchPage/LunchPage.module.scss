
.HeaderText {
  @apply text-center text-5xl font-bold mt-16 transition-all ease-in-out duration-200 hover:text-light;
}
