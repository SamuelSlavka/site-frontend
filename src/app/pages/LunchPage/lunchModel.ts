
export interface LunchState {
    lunches: LunchValue[];
    restaurants: {[id: string]: Restaurnat};
    loading: boolean;
}

export interface Restaurnat {
    endpoint: string;
    selected: boolean;
}

export interface LunchValue {
    value: string;
    restaurantId: string;
}