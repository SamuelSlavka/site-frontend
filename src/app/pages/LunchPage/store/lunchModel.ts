
export interface LunchState {
    lunches: LunchValue[];
    loading: boolean;
}

export interface LunchValue {
    id: string;
    restaurant_id: string;
    value: string[]
}