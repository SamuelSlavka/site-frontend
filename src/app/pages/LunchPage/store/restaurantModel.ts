import { EntityState } from "@reduxjs/toolkit";

export interface RestaurantState extends EntityState<RestaurantModel> {
    loading: boolean;
}

export interface RestaurantModel {
    id: string,
    restaurant_name: string,
    restaurant_description: string,
    restaurant_endpoint: string,
    selected: boolean,
}