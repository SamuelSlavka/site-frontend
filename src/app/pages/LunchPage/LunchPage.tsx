import React, { FC, useEffect } from "react";
import { useSelector } from "react-redux";
import { Link } from "react-router-dom";
import {
    fetchLunch,
    selectFilteredLunches,
    selectFilteredRestaurants,
} from "./lunchSlice";

// import styles from './LunchPage.module.scss';

import store from 'src/app/store';
import { LunchValue, Restaurnat } from "./lunchModel";
import RestaurantSelect from "./components/RestaurantSelect/RestaurantSelect";

interface LunchPageProps { }

const LunchPage: FC<LunchPageProps> = () => {
    const lunches = useSelector(selectFilteredLunches);
    const restaurans = useSelector(selectFilteredRestaurants);

    // const dispatch = useDispatch();

    const lunchList = lunches?.map((lunch: LunchValue, index: number) => (
        <section key={index} className="text-center my-2">
            {lunch.value}
        </section>
    ));

    const restaurantList = restaurans?.map((restaurant: Restaurnat, index: number) => (
        <section key={index} className="text-center my-2">
            {restaurant.endpoint}
        </section>
    ));

    useEffect(() => {
        store.dispatch(fetchLunch("newRestaurant"));
        return () => { };
    }, []);

    return (
        // className={styles.LunchPage}
        <div data-testid="LunchPage">
            <section>
                <h1 className="text-center text-5xl font-bold mt-16">Lunch</h1>
            </section>
            <section className="text-center mt-4 mb-8">
                <Link to="/">
                    <span>{"< home"}</span>
                </Link>
            </section>
            <RestaurantSelect />
            {lunchList}
            {restaurantList}
        </div>
    );
};

export default LunchPage;
