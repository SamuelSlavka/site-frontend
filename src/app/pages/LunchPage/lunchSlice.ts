import { RootState } from '@app/store';
import { LunchState, LunchValue } from './lunchModel';
import { client } from 'src/app/api/client';
import { createSelector, createSlice, PayloadAction } from '@reduxjs/toolkit';

const initialState: LunchState = {
    restaurants: {
        '01': { endpoint: '/pok', selected: true },
        '02': { endpoint: '/tus', selected: true },
        '03': { endpoint: '/ska', selected: true },
    },
    lunches: [
        { value: 'm1', restaurantId: '01' },
        { value: 'm12', restaurantId: '01' },
        { value: 'm2', restaurantId: '02' },
    ],
    loading: false,
};

export const lunchSlice = createSlice({
    name: 'lunch',
    initialState,
    reducers: {
        lunchLoaded: (state, action: PayloadAction<{ value: LunchValue }>) => {
            state.lunches.push(action.payload.value);
            state.loading = false;
        },
        lunchLoading(state) {
            state.loading = true;
        },
        toggleRestaurant: (state, action: PayloadAction<{ id: string }>) => {
            const restaurantId = action.payload.id;
            if (state.restaurants[restaurantId]) {
                state.restaurants[restaurantId].selected =
                    !state.restaurants[restaurantId].selected;
            }
        },
    },
});

export const { lunchLoaded, lunchLoading, toggleRestaurant } =
    lunchSlice.actions;
export default lunchSlice.reducer;

export const fetchLunch = (restaurantId: string) => async (dispatch: any) => {
    dispatch(lunchLoading());
    const response = await client.get('stuff');
    dispatch(lunchLoaded({ value: { value: response, restaurantId } }));
};

export const selectRestaurantEntities = (state: RootState) =>
    state.lunch.restaurants;

export const selectRestaurantById = (
    state: RootState,
    restaurantId: string
) => {
    return selectRestaurantEntities(state)[restaurantId];
};

export const selectRestaurants = createSelector(
    selectRestaurantEntities,
    (entities) => {
        const keys = Object.keys(entities);
        return keys.map((key) => {
            return { id: key, ...entities[key] };
        });
    }
);

export const selectFilteredRestaurants = createSelector(
    selectRestaurants,
    (restaurnats) => restaurnats.filter((restaurant) => restaurant.selected)
);

export const selectLunches = (state: RootState) => state.lunch.lunches;

export const selectFilteredLunches = createSelector(
    selectRestaurantEntities,
    selectLunches,
    (restaurnats, lunches) => {
        return lunches?.filter(
            (lunch) => restaurnats[lunch.restaurantId]?.selected === true
        );
    }
);
