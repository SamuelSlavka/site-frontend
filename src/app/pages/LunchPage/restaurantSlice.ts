import { RootState } from '../../store';
import { client } from '../../api/client';
import {
    createEntityAdapter, createAsyncThunk, createSelector, createSlice, PayloadAction
} from '@reduxjs/toolkit';
import { RestaurantModel, RestaurantState } from './restaurantModel';

const restaurantAdapter = createEntityAdapter<RestaurantModel>({
    selectId: (restaurant) => restaurant.id,
    sortComparer: (a, b) => a.restaurant_name.localeCompare(b.restaurant_name),
})

const initialState: RestaurantState = restaurantAdapter.getInitialState({
    loading: false,
});

export const restaurantSlice = createSlice({
    name: 'restaurant',
    initialState: initialState,
    reducers: {
        restaurantsLoaded: (state: RestaurantState, action: PayloadAction<{ restaurants: RestaurantModel[] }>) => {
            state = restaurantAdapter.upsertMany(state, action.payload.restaurants);
            state.loading = false;
        },
        restaurantsLoading(state: RestaurantState) {
            state.loading = true;
        },
        toggleRestaurant: (state: RestaurantState, action: PayloadAction<{ id: string }>) => {
            const restaurant = state.entities[action.payload.id];
            if (restaurant && action.payload.id) {
                restaurant.selected = !state.entities[action.payload.id]?.selected;
            }
        },
    },
    extraReducers: (builder) => {
        builder.addCase(fetchRestaurants.fulfilled,
            (state: RestaurantState, action: PayloadAction<RestaurantModel[]>) => {
                const selectedPayload: RestaurantModel[] = [];
                action.payload.map(restaurant => {
                    const currentRestaurant = state.entities[restaurant.id]?.selected;
                    // if the restaurant is not in store set its selection to true
                    selectedPayload.push(
                        {
                            ...restaurant,
                            selected: currentRestaurant === undefined ? true : currentRestaurant ?? false,
                        });
                    return 0;
                })
                state = restaurantAdapter.removeAll(state);
                state = restaurantAdapter.setMany(state, selectedPayload);
                state.loading = false;
            },
        )
    }
});

export const {
    selectById: selectRestaurantById,
  } = restaurantAdapter.getSelectors((state: RootState) => state.restaurant)

export const { toggleRestaurant, restaurantsLoading, restaurantsLoaded } = restaurantSlice.actions;
export default restaurantSlice.reducer;


export const fetchRestaurants = createAsyncThunk(
    'restaurants/fetch',
    async (thunkAPI) => {
        const response = await client.get('restaurants');
        return response.data
    }
)

export const selectRestaurantEntities = (state: RootState) =>
    state.restaurant.entities;


export const selectRestaurants = createSelector(
    selectRestaurantEntities,
    (entities) => {
        const keys = Object.keys(entities);
        return keys.map((key) => {
            return { id: key, ...entities[key] };
        });
    }
);

export const selectFilteredRestaurants = createSelector(
    selectRestaurants,
    (restaurants) => restaurants.filter((restaurant) => restaurant && restaurant.selected)
);