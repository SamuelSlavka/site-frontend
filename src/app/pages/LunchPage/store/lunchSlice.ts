import { RootState } from '@app/store';
import { LunchState, LunchValue } from './lunchModel';
import { createAsyncThunk, createSelector, createSlice, isAnyOf } from '@reduxjs/toolkit';
import { client } from '../../../api/client';
import { selectRestaurantEntities } from './restaurantSlice';
import { toast } from 'react-toastify';

const initialState: LunchState = {
    lunches: [],
    loading: false,
};

export const lunchSlice = createSlice({
    name: 'lunch',
    initialState,
    reducers: {},
    extraReducers: (builder) => {
        builder.addCase(fetchLunches.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(fetchLunches.fulfilled, (state, action) => {
            state.lunches = action.payload;
            state.loading = false;
        });
        builder.addCase(fetchAllLunches.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(fetchAllLunches.fulfilled, (state, action) => {
            state.lunches = action.payload;
            state.loading = false;
        });
        builder.addMatcher(isAnyOf (fetchAllLunches.rejected, fetchLunches.rejected), (state) => {
            toast.error("Load failed", { theme: "dark", autoClose: 2000, pauseOnFocusLoss: false });
            state.loading = false;
        });
    }
});

export default lunchSlice.reducer;

export const fetchLunches = createAsyncThunk(
    'lunches/fetch',
    async (restaurantList: string[]) => {
        const response = await client.post('lunches', {restaurantIds: restaurantList});
        return response.data
    }
)


export const fetchAllLunches = createAsyncThunk(
    'all_lunches/fetch',
    async () => {
        const response = await client.get('all_lunches');
        return response.data
    }
)

const selectLunches = (state: RootState) => state.lunch.lunches;

export const selectFilteredLunches = createSelector(
    selectRestaurantEntities,
    selectLunches,
    (restaurants, lunches) => {
        return lunches?.filter(
            (lunch: LunchValue) => restaurants[lunch.restaurant_id]?.selected === true
        );
    }
);
