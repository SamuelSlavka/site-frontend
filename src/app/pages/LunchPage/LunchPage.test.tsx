import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { BrowserRouter } from "react-router-dom";
import LunchPage from "./LunchPage";
import store from "src/app/store";
import { Provider } from "react-redux";

describe("LunchPage works", () => {
    test("it should mount", () => {
        render(
            <BrowserRouter>
                <Provider store={store}>
                    <LunchPage />
                </Provider>
            </BrowserRouter>
        );
        const lunchPage = screen.getByTestId("LunchPage");
        expect(lunchPage).toBeInTheDocument();
    });
});
