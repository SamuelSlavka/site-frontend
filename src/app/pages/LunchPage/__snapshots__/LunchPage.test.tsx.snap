// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LunchPage works it should render correnctly 1`] = `
<div
  className="overflow-auto HideScrollbars inline-block relative min-w-full min-h-fit h-full object-cover"
  data-testid="LunchPage"
>
  <section>
    <h1
      className="text-center pt-16"
    >
      <span
        className="HeaderText"
      >
        Lunch
      </span>
    </h1>
  </section>
  <section
    className="LinkTopContainer"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <span
        className="LinkTop"
      >
        &lt; home
      </span>
    </a>
  </section>
  <section
    className="m-4 "
  >
    <div
      className="SelectorContainer"
      data-testid="RestaurantSelect"
    >
      <div
        className="relative mt-1"
      >
        <button
          aria-expanded={false}
          aria-haspopup={true}
          className="relative w-full cursor-default rounded-lg bg-white py-2 pl-3 pr-10 text-left shadow-md focus:outline-none focus-visible:border-indigo-500 focus-visible:ring-2 focus-visible:ring-white focus-visible:ring-opacity-75 focus-visible:ring-offset-2 focus-visible:ring-offset-orange-300"
          data-headlessui-state=""
          disabled={false}
          id="headlessui-listbox-button-:r0:"
          onClick={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          type="button"
        >
          <span
            className="SearchField"
          />
          <span
            className="SearchIcon absolute inset-y-0 right-3 flex items-center pl-3 text-amber-600"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-utensils "
              data-icon="utensils"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 0C400 0 288 32 288 176V288c0 35.3 28.7 64 64 64h32V480c0 17.7 14.3 32 32 32s32-14.3 32-32V352 240 32c0-17.7-14.3-32-32-32zM64 16C64 7.8 57.9 1 49.7 .1S34.2 4.6 32.4 12.5L2.1 148.8C.7 155.1 0 161.5 0 167.9c0 45.9 35.1 83.6 80 87.7V480c0 17.7 14.3 32 32 32s32-14.3 32-32V255.6c44.9-4.1 80-41.8 80-87.7c0-6.4-.7-12.8-2.1-19.1L191.6 12.5c-1.8-8-9.3-13.3-17.4-12.4S160 7.8 160 16V150.2c0 5.4-4.4 9.8-9.8 9.8c-5.1 0-9.3-3.9-9.8-9L127.9 14.6C127.2 6.3 120.3 0 112 0s-15.2 6.3-15.9 14.6L83.7 151c-.5 5.1-4.7 9-9.8 9c-5.4 0-9.8-4.4-9.8-9.8V16zm48.3 152l-.3 0-.3 0 .3-.7 .3 .7z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </section>
  <section
    className="pt-4 pb-8 flex w-100 flex-wrap justify-center transition-all ease-in-out duration-300"
  />
</div>
`;
